src/lib/Output/Generator/AbstractFieldTypeHashGenerator.php

```diff
@@ -0,0 +1,141 @@
+
+abstract class AbstractFieldTypeHashGenerator implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    private NormalizerInterface $normalizer;
+
+    private bool $strictMode;
+
+    public function __construct(
+        NormalizerInterface $normalizer,
+        ?LoggerInterface $logger = null,
+        bool $strictMode = false
+    ) {
+        $this->normalizer = $normalizer;
+        $this->logger = $logger ?? new NullLogger();
+        $this->strictMode = $strictMode;
+    }
+
+    /**
+     * Generates the field type value $hashValue as a child of the given Object
+     * using $hashElementName as the property name.
+     */
+    public function generateHashValue(
+        DataObjectInterface $parent,
+        string $hashElementName,
+        mixed $hashValue
+    ): void {
+        $parent->$hashElementName = $this->generateValue($parent, $hashValue);
+    }
+
+    /**
+     * Generates and returns a JSON structure (array or object) depending on $value type
+     * with $parent.
+     *
+     * If $type only contains numeric keys, the resulting structure will be an
+     * JSON array, otherwise a JSON object
+     *
+     * @param array<mixed> $value
+     */
+    protected function generateArrayValue(
+        DataObjectInterface $parent,
+        array $value,
+    ): DataObjectInterface {
+        if ($this->isNumericArray($value)) {
+            return $this->generateListArray($parent, $value);
+        } else {
+            return $this->generateHashArray($parent, $value);
+        }
+    }
+
+    /**
+     * Generates and returns a value based on $hashValue type, with $parent (
+     * if the type of $hashValue supports it).
+     */
+    abstract protected function generateValue(DataObjectInterface $parent, mixed $value): mixed;
+
+    /**
+     * Checks if the given $value is a purely numeric array.
+     *
+     * @param array<mixed> $value
+     */
+    protected function isNumericArray(array $value): bool
+    {
+        foreach (array_keys($value) as $key) {
+            if (is_string($key)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    protected function generateObjectValue(DataObjectInterface $parent, object $value): mixed
+    {
+        try {
+            $value = $this->normalizer->normalize($value, 'json', ['parent' => $parent]);
+        } catch (ExceptionInterface $e) {
+            if ($this->strictMode) {
+                throw $e;
+            }
+            $message = sprintf(
+                'Unable to normalize value for type "%s". %s. '
+                . 'Ensure that a normalizer is registered with tag: "%s".',
+                get_class($value),
+                $e->getMessage(),
+                'ibexa.rest.serializer.normalizer',
+            );
+
+            assert($this->logger instanceof LoggerInterface);
+            $this->logger->error($message, [
+                'exception' => $e,
+            ]);
```

Comment:
I think the newer notation for null-coalescing operator is:
```suggestion
            $this->logger?->error($message, [
                'exception' => $e,
            ]);
```